onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -2686,6 +2686,149 @@ It's an extension of Gelu. It takes the sum of input A and bias input B as the i
         updateOutputElemType(ctx, 0, ONNX_NAMESPACE::TensorProto::BOOL);
       });
 
+  static const char* OptionalEmpty_ver1_doc = R"DOC(
+      Construct an empty optional, with given element type.
+      )DOC";
+
+  ONNX_CONTRIB_OPERATOR_SCHEMA(OptionalEmpty)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .SetDoc(OptionalEmpty_ver1_doc)
+      .Attr("type", "Type of the element in the optional output", AttributeProto::TYPE_PROTO)
+      .Output(
+          0,
+          "output",
+          "An optional that contains no elements.",
+          "O")
+      .TypeConstraint(
+          "O",
+          {"optional(tensor(float))",
+           "optional(seq(tensor(float)))"},
+          "Constrains output type to all optional tensor or optional sequence types.")
+      .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+        const auto* attr_proto = ctx.getAttribute("type");
+        if (attr_proto == nullptr)
+          fail_type_inference(
+                "Attribute type should be provided.");
+        if (!attr_proto->has_tp())
+          fail_type_inference(
+              "Attribute type should be of type TypeProto and specify a type.");
+        auto attr_tp = attr_proto->tp();
+        const size_t numOutputs = ctx.getNumOutputs();
+        if (numOutputs != 1) {
+          fail_type_inference("OptionalEmpty is expected to have 1 output.");
+        }
+        ctx.getOutputType(0)
```

Comment:
Minor nit: May be we should validate that the type is one of the two allowed by the TypeConstraint above. (In the long run, we should eliminate the type-constraint, but for now we are forced to have a TypeConstraint.)

Reply:
I am planning to add all the types in the standard spec. Kept float only here for simplicity of the POC. I will extend it.